controllers/modals/submissionPayments/SubmissionPaymentsHandler.inc.php

```diff
@@ -0,0 +1,149 @@
+		$templateMgr->assign([
+			'assignData' => [
+				'components' => [
+					FORM_SUBMISSION_PAYMENTS => $submissionPaymentsForm->getConfig(),
+				],
+			],
+		]);
+
+		return $templateMgr->fetchJson('controllers/modals/submissionPayments/status.tpl');
+
+	}
+
+	/**
+	 * Change the status of publication payments.
+	 * @param $args array
+	 * @param $request PKPRequest
+	 * @return JSONMessage JSON object
+	 */
+	public function save($args, $request) {
+		$context = $request->getContext();
+		$paymentManager = \Application::getPaymentManager($context);
+		$completedPaymentDao = \DAORegistry::getDAO('OJSCompletedPaymentDAO'); /* @var $completedPaymentDao OJSCompletedPaymentDAO */
+		$publicationFeeEnabled = $paymentManager->publicationEnabled();
+		$publicationFeePayment = $completedPaymentDao->getByAssoc(null, PAYMENT_TYPE_PUBLICATION, $this->submission->getId());
+		switch ($request->getUserVar('publicationFeeStatus')) {
+			case 'waived':
+				// Check if a waiver already exists; if so, don't do anything.
+				if ($publicationFeePayment && !$publicationFeePayment->getAmount()) break;
+
+				// If a fulfillment (nonzero amount) already exists, remove it.
+				if ($publicationFeePayment) $completedPaymentDao->deleteById($publicationFeePayment->getId());
+
+				// Record a waived payment.
+				$queuedPayment = $paymentManager->createQueuedPayment(
+					$request,
+					PAYMENT_TYPE_PUBLICATION,
+					$request->getUser()->getId(),
+					$this->submission->getId(),
+					0, '' // Zero amount, no currency
+				);
+				$paymentManager->queuePayment($queuedPayment);
+				$paymentManager->fulfillQueuedPayment($request, $queuedPayment, 'ManualPayment');
+				break;
+			case 'paid':
+				// Check if a fulfilled payment already exists; if so, don't do anything.
+				if ($publicationFeePayment && $publicationFeePayment->getAmount()) break;
+
+				// If a waiver (0 amount) already exists, remove it.
+				if ($publicationFeePayment) $completedPaymentDao->deleteById($publicationFeePayment->getId());
+
+				// Record a fulfilled payment.
+				$stageAssignmentDao = \DAORegistry::getDAO('StageAssignmentDAO'); /* @var $stageAssignmentDao StageAssignmentDAO */
+				$submitterAssignments = $stageAssignmentDao->getBySubmissionAndRoleId($this->submission->getId(), ROLE_ID_AUTHOR);
+				$submitterAssignment = $submitterAssignments->next();
+				$queuedPayment = $paymentManager->createQueuedPayment(
+					$request,
+					PAYMENT_TYPE_PUBLICATION,
+					$submitterAssignment->getUserId(),
+					$this->submission->getId(),
+					$context->getSetting('publicationFee'),
+					$context->getSetting('currency')
+				);
+				$paymentManager->queuePayment($queuedPayment);
+				$paymentManager->fulfillQueuedPayment($request, $queuedPayment, 'Waiver');
+				break;
+			case 'unpaid':
+				if ($publicationFeePayment) $completedPaymentDao->deleteById($publicationFeePayment->getId());
+				break;
+			default: throw new Exception('Unknown fee payment status!');
+		}
+	}
+}
```

Comment:
I do think that this should go into a REST API endpoint. It could either be a new `/payments/submissions/{id}` endpoint or a `/submissions/{id}/payment` endpoint. It's up to you, though.

Reply:
Do you mind if I get your help on this part? I'm a little lost in the API tooling (Swagger/Redoc) and had some trouble getting the `POST` to route to the handler function in this commit: https://github.com/pkp/ojs/pull/3018/commits/a31ea7b970c98d200e5a3435eaaf9a78986e5d9a

Otherwise everything should be fine -- it's converted from a modal to a drop-down, and no longer uses a separate request to load the form.

Reply:
(Routing works if I convert it to a `GET`, but obviously that's not what I want; I suspect there's some middleware that's not satisfied in the `POST` request, but e.g. CSRF looks fine to me.)

Reply:
Here's a PR with the fix: https://github.com/asmecher/ojs/pull/6, as well as some small adjustments to improve error handling in the API. The issue you had with `POST` is because the form uses `PUT`: https://github.com/pkp/ojs/pull/3018/files#diff-4000e872cbf9a365b8c7ce74d12caf1e79ae47c630e603c9a3616be18c95b653R25. (The `Form.vue` component will automatically convert to `POST` and use the `X-Http-Method-Override` header, but Slim still expects the route to be `PUT`.)

Can you also merge this PR, which makes some style adjustments for forms in dropdowns? https://github.com/pkp/ui-library/pull/142

Reply:
:+1: Thanks!